File: e2e/java/src/test/java/dev/kreuzberg/e2e/RobotsTest.java
package dev.kreuzberg.e2e;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import dev.kreuzberg.kreuzcrawl.Kreuzcrawl;

/** E2e tests for category: robots. */
class RobotsTest {
    @Test
    void testRobotsAllowAll() throws Exception {
        // Permissive robots.txt allows all paths
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(true, result.isAllowed());
    }

    @Test
    void testRobotsAllowOverride() throws Exception {
        // Allow directive overrides Disallow for specific paths
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(true, result.isAllowed());
    }

    @Test
    void testRobotsCommentsHandling() throws Exception {
        // Correctly parses robots.txt with inline and line comments
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(true, result.isAllowed());
    }

    @Test
    void testRobotsCrawlDelay() throws Exception {
        // Respects crawl-delay directive from robots.txt
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(2, result.crawlDelay().orElse(""));
    }

    @Test
    void testRobotsDisallowPath() throws Exception {
        // Robots.txt disallows specific paths
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(false, result.isAllowed());
    }

    @Test
    void testRobotsMetaNofollow() throws Exception {
        // Detects nofollow meta robots tag and skips link extraction
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(true, result.nofollowDetected());
    }

    @Test
    void testRobotsMetaNoindex() throws Exception {
        // Detects noindex meta robots tag in HTML page
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(true, result.noindexDetected());
    }

    @Test
    void testRobotsMissing404() throws Exception {
        // Missing robots.txt (404) allows all crawling
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(true, result.isAllowed());
    }

    @Test
    void testRobotsMultipleUserAgents() throws Exception {
        // Picks the most specific user-agent block from robots.txt
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(true, result.isAllowed());
    }

    @Test
    void testRobotsRequestRate() throws Exception {
        // Parses request-rate directive from robots.txt
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(5, result.crawlDelay().orElse(""));
        assertEquals(true, result.isAllowed());
    }

    @Test
    void testRobotsSitemapDirective() throws Exception {
        // Discovers sitemap URL from Sitemap directive in robots.txt
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(true, result.isAllowed());
    }

    @Test
    void testRobotsUserAgentSpecific() throws Exception {
        // Matches user-agent specific rules in robots.txt
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(false, result.isAllowed());
    }

    @Test
    void testRobotsWildcardPaths() throws Exception {
        // Handles wildcard Disallow patterns in robots.txt
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(false, result.isAllowed());
    }

    @Test
    void testRobotsXRobotsTag() throws Exception {
        // Respects X-Robots-Tag HTTP header directives
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals("noindex, nofollow", result.xRobotsTag().orElse(""));
        assertEquals(true, result.noindexDetected());
        assertEquals(true, result.nofollowDetected());
    }

}

File: e2e/go/robots_test.go
// E2e tests for category: robots
package e2e_test

import (
	"strings"
	"testing"

	pkg "github.com/kreuzberg-dev/kreuzcrawl"
)

func Test_RobotsAllowAll(t *testing.T) {
	// Permissive robots.txt allows all paths
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.is_allowed' not available on result type
}

func Test_RobotsAllowOverride(t *testing.T) {
	// Allow directive overrides Disallow for specific paths
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.is_allowed' not available on result type
}

func Test_RobotsCommentsHandling(t *testing.T) {
	// Correctly parses robots.txt with inline and line comments
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.is_allowed' not available on result type
}

func Test_RobotsCrawlDelay(t *testing.T) {
	// Respects crawl-delay directive from robots.txt
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.crawl_delay' not available on result type
}

func Test_RobotsDisallowPath(t *testing.T) {
	// Robots.txt disallows specific paths
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.is_allowed' not available on result type
}

func Test_RobotsMetaNofollow(t *testing.T) {
	// Detects nofollow meta robots tag and skips link extraction
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.nofollow_detected' not available on result type
}

func Test_RobotsMetaNoindex(t *testing.T) {
	// Detects noindex meta robots tag in HTML page
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.noindex_detected' not available on result type
}

func Test_RobotsMissing404(t *testing.T) {
	// Missing robots.txt (404) allows all crawling
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.is_allowed' not available on result type
}

func Test_RobotsMultipleUserAgents(t *testing.T) {
	// Picks the most specific user-agent block from robots.txt
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.is_allowed' not available on result type
}

func Test_RobotsRequestRate(t *testing.T) {
	// Parses request-rate directive from robots.txt
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.crawl_delay' not available on result type
	// skipped: field 'robots.is_allowed' not available on result type
}

func Test_RobotsSitemapDirective(t *testing.T) {
	// Discovers sitemap URL from Sitemap directive in robots.txt
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.is_allowed' not available on result type
}

func Test_RobotsUserAgentSpecific(t *testing.T) {
	// Matches user-agent specific rules in robots.txt
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.is_allowed' not available on result type
}

func Test_RobotsWildcardPaths(t *testing.T) {
	// Handles wildcard Disallow patterns in robots.txt
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.is_allowed' not available on result type
}

func Test_RobotsXRobotsTag(t *testing.T) {
	// Respects X-Robots-Tag HTTP header directives
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'robots.x_robots_tag' not available on result type
	// skipped: field 'robots.noindex_detected' not available on result type
	// skipped: field 'robots.nofollow_detected' not available on result type
}
